invenio_accounts/hash.py

```diff
@@ -0,0 +1,142 @@
+
+import hashlib
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+from flask import current_app
+from passlib.utils.compat import str_to_uascii
+from passlib.utils.handlers import HasUserContext, StaticHandler
+from six import binary_type, text_type
+
+__all__ = ('invenio_aes_encrypted_email', )
+
+
+def _to_binary(val):
+    """Convert to binary."""
+    if isinstance(val, text_type):
+        return val.encode('utf-8')
+    return val
+
+
+def _mysql_aes_key(key):
+    """Format key."""
+    final_key = bytearray(16)
+    for i, c in enumerate(key):
+        final_key[i % 16] ^= ord(key[i])
+    return bytes(final_key)
+
+
+def _mysql_aes_pad(val):
+    """Padding."""
+    pad_value = 16 - (len(val) % 16)
+    return '{0}{1}'.format(val, chr(pad_value)*pad_value)
+
+
+def _mysql_aes_unpad(val):
+    """Reverse padding."""
+    pad_value = ord(val[-1])
+    return val[:-pad_value]
+
+
+def _mysql_aes_engine(key):
+    """Create MYSQL AES cipher engine."""
+    return Cipher(algorithms.AES(key), modes.ECB(), default_backend())
+
+
+def mysql_aes_encrypt(val, key):
+    """Mysql AES encrypt value with secret key."""
+    assert isinstance(val, binary_type) or isinstance(val, text_type)
+    assert isinstance(key, binary_type) or isinstance(key, text_type)
+    k = _mysql_aes_key(_to_binary(key))
+    v = _mysql_aes_pad(_to_binary(val))
+    e = _mysql_aes_engine(k).encryptor()
+
+    return e.update(v) + e.finalize()
+
+
+def mysql_aes_decrypt(encrypted_val, key):
+    """Mysql AES decrypt value with secret key."""
+    assert isinstance(encrypted_val, binary_type) \
+        or isinstance(encrypted_val, text_type)
+    assert isinstance(key, binary_type) or isinstance(key, text_type)
+    k = _mysql_aes_key(_to_binary(key))
+    d = _mysql_aes_engine(_to_binary(k)).decryptor()
+    return _mysql_aes_unpad(d.update(_to_binary(encrypted_val)) + d.finalize())
+
+
+class InvenioAesEncryptedEmail(HasUserContext, StaticHandler):
+    """Invenio AES encryption of user email using password as secret key.
+
+    Invenio 1.x was AES encrypting the users email address with the password
+    as the secret key and storing it in a blob column. This e.g. caused
+    problems when a user wanted to change email address.
+    This hashing engine, differs from Invenio 1.x in that it sha256 hashes the
+    encrypted value as well to produce a string in the same length instead of
+    a binary blob. It is not done for extra security, just for convenience of
+    migration to using passlib's sha512.
+    An upgrade recipe is provided to migrated existing binary password hashes
+    to hashes of this engine.
+    """
+
+    name = "invenio_aes_encrypted_email"
+
+    def _calc_checksum(self, secret):
+        """Calculate string."""
+        return str_to_uascii(
+            hashlib.sha256(mysql_aes_encrypt(self.user, secret)).hexdigest()
+        )
+
+
+def migrate_hash(user):
```

Comment:
Well, there is a problem with the `user` context keyword that the legacy scheme uses. `flask-security`'s built-in methods `verify_password` and `encrypt_password` do not accept context arguments, therefore there is no way to not interfere with the internal `_pwd_context`. Maybe I am missing something...